base/linalg/rowvector.jl

```diff
@@ -0,0 +1,214 @@
+@pure check_types{T1,T2}(::Type{T1},::Type{T2}) = T1 === transpose_type(T2) ? nothing : error("Element type mismatch. Tried to create a `RowVector{$T1}` from an `AbstractVector{$T2}`")
+
+# The element type may be transformed as transpose is recursive
+@inline transpose_type{T}(::Type{T}) = promote_op(transpose, T)
+
+# Constructors that take a vector
+@inline RowVector{T}(vec::AbstractVector{T}) = RowVector{transpose_type(T),typeof(vec)}(vec)
+@inline (::Type{RowVector{T}}){T}(vec::AbstractVector{T}) = RowVector{T,typeof(vec)}(vec)
+
+# Constructors that take a size and default to Array
+@inline (::Type{RowVector{T}}){T}(n::Int) = RowVector{T}(Vector{transpose_type(T)}(n))
+@inline (::Type{RowVector{T}}){T}(n1::Int, n2::Int) = n1 == 1 ? RowVector{T}(Vector{transpose_type(T)}(n2)) : error("RowVector expects 1×N size, got ($n1,$n2)")
+@inline (::Type{RowVector{T}}){T}(n::Tuple{Int}) = RowVector{T}(Vector{transpose_type(T)}(n[1]))
+@inline (::Type{RowVector{T}}){T}(n::Tuple{Int,Int}) = n[1] == 1 ? RowVector{T}(Vector{transpose_type(T)}(n[2])) : error("RowVector expects 1×N size, got $n")
+
+# Conversion of underlying storage
+convert{T,V<:AbstractVector}(::Type{RowVector{T,V}}, rowvec::RowVector) = RowVector{T,V}(convert(V,rowvec.vec))
+
+# similar()
+@inline similar(rowvec::RowVector) = RowVector(similar(rowvec.vec))
+@inline similar{T}(rowvec::RowVector, ::Type{T}) = RowVector(similar(rowvec.vec, transpose_type(T)))
+# There is no resizing similar() because it would be ambiguous if the result were a Matrix or a RowVector
+
+# Basic methods
+"""
+    transpose(v::AbstractVector)
+
+The transposition operator (`.'`).
+
+# Example
+
+```jldoctest
+julia> v = [1,2,3]
+3-element Array{Int64,1}:
+ 1
+ 2
+ 3
+
+julia> transpose(v)
+1×3 RowVector{Int64,Array{Int64,1}}:
+ 1  2  3
+```
+"""
+@inline transpose(vec::AbstractVector) = RowVector(vec)
+@inline ctranspose{T}(vec::AbstractVector{T}) = RowVector(conj(vec))
+@inline ctranspose{T<:Real}(vec::AbstractVector{T}) = RowVector(vec)
+
+@inline transpose(rowvec::RowVector) = rowvec.vec
+@inline ctranspose{T}(rowvec::RowVector{T}) = conj(rowvec.vec)
+@inline ctranspose{T<:Real}(rowvec::RowVector{T}) = rowvec.vec
+
+parent(rowvec::RowVector) = rowvec.vec
+
+# Strictly, these are unnecessary but will make things stabler if we introduce
+# a "view" for conj(::AbstractArray)
+@inline conj(rowvec::RowVector) = RowVector(conj(rowvec.vec))
+@inline conj{T<:Real}(rowvec::RowVector{T}) = rowvec
+
+# AbstractArray interface
+@inline length(rowvec::RowVector) =  length(rowvec.vec)
+@inline size(rowvec::RowVector) = (1, length(rowvec.vec))
+@inline size(rowvec::RowVector, d) = ifelse(d==2, length(rowvec.vec), 1)
+@inline indices(rowvec::RowVector) = (Base.OneTo(1), indices(rowvec.vec)[1])
+@inline indices(rowvec::RowVector, d) = ifelse(d == 2, indices(rowvec.vec)[1], Base.OneTo(1))
+linearindexing{V<:RowVector}(::Union{V,Type{V}}) = LinearFast()
```

Comment:
Right, this is something that'd be nice to change at some point.  LinearFast (read: index-with-one-dimension) and LinearSlow (read: index-with-N-dimensions) represent the same thing when N=1.  It'd be worth seeing if the performance of SparseVector changes if you make it `LinearFast` — it may end up hitting simpler methods that way.  If so, we could change the default for all vectors.  And maybe eventually re-work this to be clearer (https://github.com/JuliaLang/julia/issues/20175#issuecomment-274349277).